packages/lesswrong/server/sql/SqlFragment.ts

```diff
@@ -54,31 +54,138 @@ export const getResolverCollection = (
+      const dependency = match[1];
+      if (fragments[dependency]) { // Only include dependencies that exist in our fragments
+        dependencies.push(dependency);
+      }
+    }
+    
+    graph[fragmentName] = dependencies;
+    visited[fragmentName] = false;
+  }
+  
+  // Depth-first search for topological sorting
+  function dfs(node: string): void {
+    // Check for cycles
+    if (temp[node]) {
+      throw new Error(`Circular dependency detected in fragments involving "${node}"`);
+    }
+    
+    if (!visited[node]) {
+      temp[node] = true; // Mark as being processed
+      
+      // Process dependencies first
+      for (const dependency of graph[node]) {
+        dfs(dependency);
+      }
+      
+      visited[node] = true;
+      temp[node] = false; // Done processing
+      result.push(node);
+    }
+  }
+  
+  // Run DFS for all fragments
+  for (const fragmentName in fragments) {
+    if (!visited[fragmentName]) {
+      dfs(fragmentName);
+    }
+  }
+  
+  // Return in proper dependency order (dependencies first), with the fragment name and its dependencies
+  return result.map(name => [name, graph[name]]);
+}
+
 /**
  * `SqlFragment` contains the logic for parsing GraphQL fragments into a
  * form that can be efficiently converted into SQL projections. As this is
- * relatively expensive, each fragment is parsed only once when it is first
- * requested and the value is then memoized globally (see `fragments.ts`).
+ * relatively expensive, each fragment group is parsed only once when it is first
+ * requested and the value is then memoized globally (see `allFragments.ts`).
  *
  * The man external interface here is the `buildProjection` method that
  * generates a `ProjectionContext` corresponding to a specific query with
  * specific arguments.
  */
-class SqlFragment {
+export class SqlFragment {
   private lexer: FragmentLexer;
   private parsedEntries: SqlFragmentEntryMap | undefined;
+  private sqlFragmentsByName: Record<string, SqlFragment> = {};
 
   constructor(
-    /** The raw source text of the fragment */
+    /** The name of the fragment that represents the result of this query */
+    resultFragmentName: string,
+    /** The raw source text of the fragment(s), including any subfragments used by the result fragment */
     fragmentSrc: string,
-    /**
-     * A function that takes a fragment name and returns the registered SQL
-     * fragment. This is taken as an argument to avoid an import cycle (DI to
-     * the rescue).
-     */
-    private getFragment: (name: FragmentName) => SqlFragment | null,
   ) {
-    this.lexer = new FragmentLexer(fragmentSrc);
+    // 1. Extract all the fragments from the source text
+    const fragmentTextByName = extractFragmentDefinitions(fragmentSrc);
+
+    // 2. Run a topological sort on the fragments to get a dependency order based on what subfragments are referenced
+    const sortedFragmentNames = topologicalSort(fragmentTextByName);
```

Comment:
Could you maybe add tests for these two functions (`extractFragmentDefinitions`, and `topologicalSort`)? I know it's generally an anti-pattern to test functions that don't otherwise need to be exported, but given that a lot of the complexity lives in these two I think it would be worthwhile

Reply:
will do!